File: ms-appointment/app/infrastructure/messaging/kafka_event_publisher.py
import json
import logging
from typing import Any, Dict, Iterable, Optional

from kafka import KafkaProducer

from ...domain.events import (
    AppointmentCancelled,
    AppointmentCreated,
    DomainEvent,
)
from ...domain.ports.event_publisher import EventPublisherPort

log = logging.getLogger("DomainEvents")

TOPIC = "appointment-events"


class KafkaEventPublisher(EventPublisherPort):
    """Publica domain events a Kafka. Preserva el contrato de notification_payload:
    - appointment.created envía el payload completo (con patient_email/doctor_email)
    - appointment.cancelled mezcla notification_payload si se pasa
    Otros domain events se loguean y se publican con sus campos propios.
    """

    def __init__(self, producer: "KafkaProducer") -> None:
        self._producer = producer

    def publish_all(
        self,
        events: Iterable[DomainEvent],
        notification_payload: Optional[Dict[str, Any]] = None,
    ) -> None:
        for event in events:
            log.info(
                "%s @ %s :: %s",
                event.event_name(),
                event.occurred_at.isoformat(),
                event,
            )

            aggregate_id = (
                getattr(event, "appointment_id", None)
                or getattr(event, "time_block_id", None)
                or "unknown"
            )

            if isinstance(event, AppointmentCreated):
                data = notification_payload or {
                    "appointment_id": event.appointment_id,
                    "patient_id": event.patient_id,
                    "doctor_id": event.doctor_id,
                    "type": event.care_type,
                }
            elif isinstance(event, AppointmentCancelled):
                data = {"appointment_id": event.appointment_id}
                if notification_payload:
                    data.update(notification_payload)
            else:
                data = {
                    k: v
                    for k, v in event.__dict__.items()
                    if not k.startswith("_")
                }

            envelope = {
                "eventName": event.event_name(),
                "occurredAt": event.occurred_at.isoformat(),
                "aggregateId": aggregate_id,
                "version": 1,
                "data": data,
            }
            log.info(
                "%s -> %s key=%s",
                event.event_name(),
                TOPIC,
                aggregate_id,
            )
            self._producer.send(
                TOPIC,
                key=aggregate_id.encode("utf-8"),
                value=json.dumps(envelope).encode("utf-8"),
            )

        # flush síncrono: garantiza entrega antes de salir del UoW
        self._producer.flush()

File: ms-auth/src/auth/infrastructure/messaging/kafka-event-publisher.ts
import { Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { Kafka, Producer } from 'kafkajs';
import { DomainEvent } from '../../../../shared/domain-event';
import { EventPublisher } from '../../domain/ports/out/event-publisher';

const TOPIC = 'auth-events';

@Injectable()
export class KafkaEventPublisher
  implements EventPublisher, OnModuleInit, OnModuleDestroy
{
  private readonly logger = new Logger('DomainEvents');
  private readonly producer: Producer;

  constructor() {
    const kafka = new Kafka({
      clientId: 'ms-auth',
      brokers: [(process.env.KAFKA_BOOTSTRAP_SERVERS || 'localhost:9092')],
    });
    this.producer = kafka.producer();
  }

  async onModuleInit(): Promise<void> {
    await this.producer.connect();
    this.logger.log('Kafka producer conectado (auth-events)');
  }

  async onModuleDestroy(): Promise<void> {
    await this.producer.disconnect();
  }

  async publishAll(events: DomainEvent[]): Promise<void> {
    if (!events?.length) return;

    const messages = events.map((event) => {
      const e = event as Record<string, unknown>;
      const aggregateId = String(e['userId'] ?? e['email'] ?? 'unknown');
      const envelope = {
        eventName: event.eventName(),
        occurredAt: event.occurredAt.toISOString(),
        aggregateId,
        version: 1,
        data: { ...event },
      };
      this.logger.log(
        `${envelope.eventName} @ ${envelope.occurredAt} -> ${TOPIC} key=${aggregateId}`,
      );
      return { key: aggregateId, value: JSON.stringify(envelope) };
    });

    await this.producer.send({ topic: TOPIC, messages });
  }
}
